sympy/simplify/cse_main.py

```diff
@@ -416,12 +416,23 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
     reduced_exprs : list of sympy expressions
         The reduced expressions with all of the replacements above.
     """
-    from sympy.matrices import Matrix
+    from sympy.matrices import Matrix, SparseMatrix
 
     # Handle the case if just one expression was passed.
     if isinstance(exprs, Basic):
         exprs = [exprs]
 
+    copy = exprs
+    temp = []
+    for e in exprs:
+        if isinstance(e, Matrix) or isinstance(e, SparseMatrix):
```

Comment:
cse should not treat SparseMatrix like a Matrix and generate potentially *many* zeros. Please see the code example that I gave where the pertinent data structures for each are used. Feel free to ask questions, too.